forge-game/src/main/java/forge/game/GameSnapshot.java

```diff
@@ -0,0 +1,511 @@
+                toGame.getStack().add(newSa);
+            }
+        }
+    }
+
+    public void copyGameState(Game fromGame, Game toGame, boolean restore) {
+        toGame.setAge(fromGame.getAge());
+        toGame.dangerouslySetTimestamp(fromGame.getTimestamp());
+
+        // TODO countersAddedThisTurn
+
+        // I think this is slightly wrong. I Don't think we need player maps
+        if (fromGame.getStartingPlayer() != null) {
+            toGame.setStartingPlayer(findBy(toGame, fromGame.getStartingPlayer()));
+        }
+        if (fromGame.getMonarch() != null) {
+            toGame.setMonarch(findBy(toGame, fromGame.getMonarch()));
+        }
+        if (fromGame.getMonarchBeginTurn() != null) {
+            toGame.setMonarchBeginTurn(findBy(toGame, fromGame.getMonarchBeginTurn()));
+        }
+        if (fromGame.getHasInitiative() != null) {
+            toGame.setHasInitiative(findBy(toGame, fromGame.getHasInitiative()));
+        }
+        if (fromGame.getDayTime() != null) {
+            toGame.setDayTime(fromGame.getDayTime());
+        }
+
+        for(Card fromCard : fromGame.getCardsInGame()) {
+            Card newCard = toGame.findById(fromCard.getId());
+            Player toPlayer = findBy(toGame, fromCard.getController());
+            ZoneType fromType = fromCard.getZone().getZoneType();
+
+            if (newCard == null) {
+                // Storing a game uses this path...
+                newCard = createCardCopy(toGame, toPlayer, fromCard);
+            } else {
+                ZoneType type = newCard.getZone().getZoneType();
+                if (type != fromType) {
+                    if (type.equals(ZoneType.Stack)) {
+                        toGame.getStackZone().remove(newCard);
+                    } else {
+                        toPlayer.getZone(type).remove(newCard);
+                    }
+                }
+            }
+
+            if (fromType.equals(ZoneType.Stack)) {
+                toGame.getStackZone().add(newCard);
+            } else {
+                toPlayer.getZone(fromType).add(newCard);
+            }
+
+            // TODO: This is a bit of a mess. We should probably have a method to copy a card's state.
+            newCard.setGameTimestamp(fromCard.getGameTimestamp());
+            newCard.setLayerTimestamp(fromCard.getLayerTimestamp());
+            newCard.setZone(fromCard.getZone());
+            newCard.setTapped(fromCard.isTapped());
+            newCard.setFaceDown(fromCard.isFaceDown());
+            newCard.setManifested(fromCard.isManifested());
+            newCard.setSickness(fromCard.hasSickness());
+            newCard.setState(fromCard.getCurrentStateName(), false);
+        }
+
+        // This loop happens later to make sure all cards are in the correct zone first
+        for (Card newCard : toGame.getCardsIn(ZoneType.Battlefield)) {
+            Card fromCard = fromGame.findById(newCard.getId());
+
+            if (fromCard.isAttachedToEntity()) {
```

Comment:
Got a repro scenario? I haven't seen this one.

Reply:
I just tried to undo _Lithoform Engine_ payment 🤷‍♂️

Reply:
Hmm.. not crashing for me... Some other weirdness, but no crash. What were you targeting?

Reply:
Just a good old Bear Cub 🤔 
Now that I try again however it just puts the ability on the stack even if I cancel mana payment
I'll test this further after this progresses

Reply:
Yea that was the weirdness i was seeing. Which is weird, everything else had been rolling back just fine.

Reply:
Ooohhh its not adding it anyway. Its repopulating the stack incorrectly. Shouldn't be too bad to fix.

Reply:
Ok @tool4ever I think this should be resolved. Manapool is fixed too, but I need to do some normalizing of the data since the mana is pointing to the wrong game underneath right now. I think things should work for the most part.

Reply:
Maybe it's limited to GameStates?
But this scenario related to my other concern seems to always crash for me:
tap Doubling Cube -> tap Swamp -> Activate and then cancel the Altar

[Uploading undo_test.txt…]()

Reply:
Yea ok, that was it. Now I'm stashing the snapshot after loading a game state. So this should be fine now.